applications/asset_tracker/src/main.c

```diff
@@ -812,6 +860,42 @@ static void app_connect(struct k_work *work)
 	}
 }
 
+/**@brief Disconnects from cloud. First it tries using the cloud backend's
+ *	  disconnect() implementation. If that fails, it falls back to close the
+ *	  socket directly, using close().
+ */
+static void app_disconnect(void)
+{
+	int err;
+
+	atomic_set(&send_data_enable, 0);
+	printk("Disconnecting from cloud.\n");
+
+	err = cloud_disconnect(cloud_backend);
+	if (err == -ENOTCONN) {
+		printk("Cloud connection was not established.\n");
+		return;
+	}
+
+	if (err) {
+		printk("Could not disconnect from cloud, err: %d\n", err);
+		printk("Closing the cloud socket directly\n");
+
+		err = close(cloud_backend->config->socket);
```

Comment:
If we close the socket manually, is it correct to block on the semaphore below? I wouldn't expect any cloud event to take place in that case.

Reply:
You're right, the cloud's `fd` wouldn't exist anymore after that, so no events would occur on the socket. Not sure what happens when the `poll` times out when polling a non-existent `fd`. Anyway, that could take ages and be unpredictable, just returning directly here instead.